### DISCUSSION TITLE
What happens when I press save?

### DISCUSSION THREAD

**Discussion by sglyon** [jupyterlite/jupyterlite#887] [Q&A] [answered]
I’m testing to see whether I can embed jupyterlite in an application I'm building. In order to do this, I would need be able to hook into or customize the behavior of the "save notebook" functionality. 

I've done some digging and have learned the following:

- The actual file saving into IndexedDB (by default on my browser) seems to be happening here: https://github.com/jupyterlite/jupyterlite/blob/c4360950ff7341107f7f3b4ec0502b99485254d8/packages/contents/src/contents.ts#L459
- This `IContents.save` method could be called in the router when a PUT request is made to `/api/contents/*`: https://github.com/jupyterlite/jupyterlite/blob/c4360950ff7341107f7f3b4ec0502b99485254d8/packages/server-extension/src/index.ts#L180-L188, but I can't find a place in the code where that PUT request is being made.

What I'm struggling to uncover is what actually happens when a user clicks the floppy disk icon, does cmd+s, or file > Save Notebook fire off the `IContents.save` function call. It would be perfect for me if I could somehow extend the file save logic.

**Comment by jtpio** [comment, marked as answer]
Thanks @sglyon for starting this discussion.

When a user saves a notebook, a request is sent to the server. However in JupyterLite there is no usual Python server. However there is an in-browser "server" handling these requests. This is done by patching the `WebSocket` and `fetch` from the `ServerSettings` so the requests are properly redirected:

https://github.com/jupyterlite/jupyterlite/blob/c4360950ff7341107f7f3b4ec0502b99485254d8/packages/server/src/app.ts#L27-L31

Then the requests to manage the content are routed to the places you linked above.

> It would be perfect for me if I could somehow extend the file save logic.

Another way to extend this logic would be to create a custom drive for JupyterLab (which would work in JupyterLite). A custom drive shows up as a different panel in the left area. See this extension as an example: https://github.com/jupyterlab-contrib/jupyterlab-filesystem-access

This would allow completely bypassing the default contents manager. But you would also have to implement the other methods of the `IDrive` interface.

Another way would be to provide a custom plugin for the server contents to replace the default one: 

https://github.com/jupyterlite/jupyterlite/blob/c4360950ff7341107f7f3b4ec0502b99485254d8/packages/server-extension/src/index.ts

For example by extending `Contents` and providing a custom `save` method. And providing this new plugin via a server extension: https://jupyterlite.readthedocs.io/en/latest/howto/extensions/server.html

**Reply by sglyon** [reply]
Very helpful, thank you!

**Comment by sglyon** [comment]
Thanks for the tip @jtpio 

I've tried to make my own server extension over here: https://github.com/sglyon/jupyteach-jupyterlite-contents

If I bring that extension into the jupyterlite developer setup I have and swap out the IContents provider with that one, I can see that it works (files are still saved and I see my console.log show up). All I needed to do was change `Contents` on [this line](https://github.com/jupyterlite/jupyterlite/blob/52ba29fddd7cbf45451e8235f5a1fac81dc07d29/packages/server-extension/src/index.ts#L80) to the new class I created. This is great 👏 

Next, I want to be able to add this on to replace the IContents provider without me having to go in and patch the `server-extension`. I was hoping to be able to do this by (1) installing my package via pip (2) creating the following `jupyter-lite.json` config file (see below) and (3) running `jupyter lite build`

```js
{
    "jupyter-lite-schema-version": 0,
    "jupyter-config-data": {
        "disabledExtensions": [
            "@jupyterlite/server-extension:contents"
        ]
    }
}
```

When I go this route I can see my custom package is activated (`activating custom contents plugin` shows up in console logs), but I see an error from the contents-routes plugin: 

```
Plugin '@jupyterlite/server-extension:contents-routes' failed to activate.
Error: No provider for: @jupyterlite/contents:IContents.
```

It seems like the `contents-routes` plugin defined in `packages/server-extension/src/index.ts` is trying to load _before_ my extension that provides IContents. 

- - -

So, I next attempted to just not disable the `"@jupyterlite/server-extension:contents"` extension in the `jupyter-lite.json` config file. When I do this and build jupyter lite I no longer get any complaints about no `IContents` provider, my extension continues to load (still see it's initialization console log), but my extension is never actually called. When I go to save a file the original `IContents` provider defined in `packages/contents` is used instead of mine. 

Do you know of any ways I can fix this?

**Reply by jtpio** [reply]
I think your approach looks good. Thanks for providing detailed information :+1: 
 
> It seems like the `contents-routes` plugin defined in `packages/server-extension/src/index.ts` is trying to load _before_ my extension that provides IContents.

You custom plugin should look like the default one, which provides `IContents`: 

https://github.com/jupyterlite/jupyterlite/blob/52ba29fddd7cbf45451e8235f5a1fac81dc07d29/packages/server-extension/src/index.ts#L66-L88

You may have to add this configuration to the `package.json` as well, similar to what the custom kernels do: https://github.com/jupyterlite/xeus-python-kernel/blob/62462d164c009805a0725ed11853c493dafc894b/package.json#L91-L96

```json
"sharedPackages": {
  "@jupyterlite/contents": {
  "bundled": false,
  "singleton": true
  }
}
```

**Reply by sglyon** [reply]
So I did copy/paste exactly that plugin from `server-extension`. The one I wrote does provide `IContents` (AFAICT) You can see it here:

https://github.com/sglyon/jupyteach-jupyterlite-contents/blob/36e3c9dc920f9e4228b5c8ecfa9cbf7c55714d10/src/index.ts#L10-L30

The only changes I made were (1)  swap out the class that is instantiated to be `new JupyteachContents` instead of `new Contents` (2) add a `console.log` so I know when the extension is activated and (3) change the plugin `id`

I'm not totally sure what extra fields to add to package.json. I will do some investigating, but if you could give me any tips/pointers that would be great!

**Reply by bollwyvl** [reply]
The above `sharedPackages` goes under `package.json#/jupyterlab/sharedPackages`. This is needed to ensure you provide _exactly_ the same `Token` that everything else is expecting. This is explained in greater detail on the upstream [JupyterLab Extension Developer Guide](https://jupyterlab.readthedocs.io/en/stable/extension/extension_dev.html#deduplication-of-dependencies).

**Reply by jtpio** [reply]
@sglyon I opened https://github.com/sglyon/jupyteach-jupyterlite-contents/pull/1 with the above suggestion, although not tested, if you want to have a look.

**Reply by sglyon** [reply]
Awesome thanks @bollwyvl and @jtpio. Making progress...

Now I see this in the console:

```
Failed to create module: package: @jupyteach/jupyterlite-contents; module: ./extension 
Error: Shared module @jupyterlite/contents doesn't exist in shared scope default
````

**Reply by jtpio** [reply]
Ah, I was wondering why it would work for `@jupyterlite/kernel` but not `@jupyterlite/contents`.

This is likely because `@jupyterlite/contents` (and some other packages)  is missing from the `singletonPackages` here: 

https://github.com/jupyterlite/jupyterlite/blob/52ba29fddd7cbf45451e8235f5a1fac81dc07d29/app/lab/package.json#L235-L239

This would need to be fixed in JupyterLite and released so it can be consumed by extensions.

**Reply by sglyon** [reply]
Ahh ok, thanks for the heads up. I'm guessing the solution here is to first make some changes in this repo?

I was also doing some digging and noticed that `@jupyterlite/contents` defines two tokens:

https://github.com/jupyterlite/jupyterlite/blob/52ba29fddd7cbf45451e8235f5a1fac81dc07d29/packages/contents/src/tokens.ts#L13

and 

https://github.com/jupyterlite/jupyterlite/blob/52ba29fddd7cbf45451e8235f5a1fac81dc07d29/packages/contents/src/tokens.ts#L194-L196

I only plan to offer a custom version of the  first one. Is that a problem? Do I somehow need to specify that I'm providing `IContents` in the package.json, or only in the actual specification of the plugin?

**Reply by jtpio** [reply]
> I'm guessing the solution here is to first make some changes in this repo?

Yes. If you feel like opening a PR (even as a draft) to add the missing packages that would be great, thanks!

I was planning to make a new release soon since a couple of enhancement PRs were merged recently. We could get that change in first.

> Do I somehow need to specify that I'm providing IContents in the package.json, or only in the actual specification of the plugin?

You can provide tokens individually so there shouldn't be any need to provide `IBroadcastChannelWrapper` in your extension.

**Reply by sglyon** [reply]
ok great, I did just try to open a PR. https://github.com/jupyterlite/jupyterlite/pull/917

I basically added just that one line. I will be doing some homework to see if there are other things I need to do and will also watch that PR for additional comments. 

Thanks for all the help!
